### PR TITLE
Snail mail feature branch

### PR DESCRIPTION
**Pull request by blackmad**
Hi all, 

This is the WORK IN PROGRESS version of integrating mail-your-rep into defund12.org

I think were we to release this quickly, I'd take out the entries in EmailList, so these could only be deep linked to, test that for a day or two, and then consider integrating it into the main list?

I still need to
- test that it all works (filling in the template is broken, about to fix that)
- make it easier/work to specify "officials" (I am debating between just using the google roles/levels: https://developers.google.com/civic-information/docs/v2/representatives or abstracting that into something simpler like citycouncil | mayor | statelegislator | governor | us-senator | us-representative | vice-president | president
- move stripe keys to site config
- consider moving the backend functions to defund12 also (I'd rather focus on frontend for now and move the backends separately - they're CORS'd such that any frontend can use them)
- probably a bunch of style and file naming clean up - some things currently called EmailX might need to change

notes: 
- the google API key in here is request scoped for defund12.org and I'm happy to pay if it exceeds free usage tier

to be very clear on how $$ works
- it costs $0.70 to send a one page letter, $0.80 to send a two page letter.
- it costs $0.33 to process a credit card transaction

so on every one page letter I lose 3 cents, on a letter with a long body I lose 13 cents, on a user sending two one page letters in one go I make 27 cents (each letter costs me 0.70, only pay transaction fee once - $1.73 vs $2). I am not trying to make money here, I am trying to keep the $ logic simple and easy. If they sent two two page letters I would make 7 cents ... seems like this will all be a wash ...

right now this is on the mail email list with different emojis:
image
and the form kinda looks like this
image

I think the most helpful feedback now would be
- how do people feel about the template structure
- how would people like to specify 
- how far is this PR from something that you'd consider merging / what major work do it would need
- major code/style critiques

### CODE REVIEW

**Comment by blackmad** [comment]
This is now working for me to send letters in TEST (no payment, no actual sending) from localhost

To test the test payment flow,  you can use the credit card number "4242 4242 4242 4242" with any other information around it

File: src/components/letter/LetterForm.tsx [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+
+  if (!emailKey) {
+    variables = [...variables, "YOUR EMAIL"];
+    emailKey = "YOUR EMAIL";
+  }
+
+  // When the inputted address changes, if it is fully specified
+  // geocode it and search for representatives' addresses
+  useEffect(() => {
+    if (
```
Comment (Ceshion) [inline src/components/letter/LetterForm.tsx:147]: factor this and the handlers into their own functions?

File: src/components/letter/representative-apis.tsx [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+      link = cityCouncilMember.urls?.[0];
+    }
+
+    return cityCouncilMember.addresses.map((address) => {
+      const officeName = address.name
+        ? cityCouncilMember.office.name + " - " + address.name
+        : cityCouncilMember.office.name;
+      return {
+        address: {
+          name: cityCouncilMember.name,
```
Comment (Ceshion) [inline src/components/letter/representative-apis.tsx:51]: maybe we can construct these rather than building them as literals?

File: src/components/letter/representative-apis.tsx [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+          district = office.divisionId.match(districtPattern)?.[1];
+        }
+
+        const official = reps.officials[officialIndex];
+        if (!official.address || official.address.length === 0) {
+          return undefined;
+        }
+        const address = official.address[0];
+        return {
+          address: {
```
Comment (Ceshion) [inline src/components/letter/representative-apis.tsx:144]: construct this as well?

File: src/components/letter/representative-apis.tsx [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+  // One to google, which has perfect information as far as I can tell for mayorial level and up everywhere
+  // Second to blackmad's city council API which tries to fill holes in major cities where google's city council coverage is lacking (SF, NYC)
+  let googlePromise: Promise<OfficialAddress[]> = Promise.resolve([]);
+  if (!cityCouncilOnly) {
+    googlePromise = fetchRepsFromGoogle({ address, googleApiKey, restricts });
+  }
+
+  const blackmadPromise = fetchRepsFromBlackmad({ address, restricts });
+
+  const googleResponse = await googlePromise;
```
Comment (Ceshion) [inline src/components/letter/representative-apis.tsx:252]: `Promise.all` these?

File: src/components/letter/types.ts [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+  divisionId: string;
+  officialIndices: number[];
+}
+
+export interface GoogleCivicRepsResponse {
+  officials: GoogleCivicRepsResponseOfficial[];
+  offices: GoogleCivicRepsResponseOffice[];
+}
+
+export interface BlackmadCityCountilResponseOfficial {
```
Comment (Ceshion) [inline src/components/letter/types.ts:88]: typo here, `Countil` > `Council`

File: gatsby-node.js [hunk 6]
```diff
@@ -1,44 +1,56 @@
...
-    {
-      allMarkdownRemark(filter: { frontmatter: { layout: { eq: "email" } } }) {
-        edges {
-          node {
-            frontmatter {
-              permalink
-              redirect_from
+  /** Shared logic between creating email and letter pages */
+  async function createPagesHelper({layout, component}) {
+    const emailData = await graphql(`
```
[the remark this answers is not in the document]
Comment (blackmad) [reply gatsby-node.js:6]: renamed to createEmailOrLetterTemplatePage, renamed emailData to templateData

this file didn't have any types because it was .js, I converted it to .ts wooo

File: src/components/letter/CheckoutForm.tsx [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+  let buttonText = "";
+  if (inSubmit) {
+    buttonText = "Submitting ...";
+  } else if (
+    myAddress.address_line1 &&
+    checkedAddresses.length === 0 &&
+    !formValid
+  ) {
+    buttonText = "Select some addresses and fill in all fields";
+  } else if (myAddress.address_line1 && checkedAddresses.length === 0) {
```
[the remark this answers is not in the document]
Comment (blackmad) [reply src/components/letter/CheckoutForm.tsx:96]: moved this to a function to be more readable

File: src/components/letter/CheckoutForm.tsx [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+  email,
+  variables,
+}: {
+  checkedAddresses: Address[];
+  myAddress: Address;
+  body: string;
+  formValid: boolean;
+  email: string;
+  variables: Record<string, string>;
+}): ReactElement => {
```
[the remark this answers is not in the document]
Comment (blackmad) [reply src/components/letter/CheckoutForm.tsx:33]: I don't really see a clear way to break this down further that doesn't do a ton of prop-drilling/passing to get there which I don't think helps right now.

I moved totalAmount into the makeButtonText function, buttonText is no longer a variable, isDisabled is still there.

I'm not sure what you mean about refactoring out the initialization logic ... the useState() bits are just what new functional react components look like. And hooks really should be at the top of the function (because it's a requirement that you always have the same hooks generated in a functional component - if you don't put all that at the top, you run the risk of having an early return that means you're making fewer hooks on some runs than others)

File: src/components/letter/Letter.tsx [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+  }
+
+  /**
+   * React render method.
+   * @return {React.ReactNode} the rendered component
+   */
+  render(): React.ReactNode {
+    // parse the offical restricts from the template markdown that look like
+    // locality:legislatorUpperBody into structured restricts for our APIs
+    const officialRestricts = this.letterData.officials?.map(
```
[the remark this answers is not in the document]
Comment (blackmad) [reply src/components/letter/Letter.tsx:38]: That's not the react style I'm used to. I'm used to only putting initializations in componentDidMount that are heavy, and everything else that's setting up the logic you need for render ... goes in render. Also, componentDidMount is going away with functional components (it becomes "useEffect" which is pretty clear it's meant for side-effects, not calculations).

It's only going to be called once because the props that are being passed in aren't changing so it's not a performance issue

File: src/components/letter/LetterForm.tsx [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+        })
+      : officials;
+
+  if (myAddress.address_line1 && officialAddresses.length === 0) {
+    return <div>No representatives found, sorry</div>;
+  }
+
+  return (
+    <div className="pure-controls">
+      {officialAddresses?.map((officialAddress) => {
```
[the remark this answers is not in the document]
Comment (blackmad) [reply src/components/letter/LetterForm.tsx:56]: moved into its own component

File: src/components/letter/types.ts (hunk 5 above)
Comment (blackmad) [reply src/components/letter/types.ts:88]: fixed everywhere

File: src/components/letter/representative-apis.tsx (hunk 4 above)
Comment (blackmad) [reply src/components/letter/representative-apis.tsx:252]: they're being used individually in the following logic - what would a Promise.all help with here?

const [googleResponse, blackmadResponse] = await Promise.all([googlePromise, blackmadPromise])? ... this is less good to me because you would run the risk of getting the inputs out of order

File: src/components/letter/LetterForm.tsx [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+      })}
+    </div>
+  );
+}
+
+/** Renders all the input fields to fill in the letter and complete the transaction
+ *
+ * @return {React.ReactNode} the rendered component
+ */
+function Inputs({
```
[the remark this answers is not in the document]
Comment (blackmad) [reply src/components/letter/LetterForm.tsx:86]: all moved to their own files + given better names

File: src/components/letter/LetterForm.tsx [hunk 12]
```diff
@@ -0,0 +1,283 @@
...
+  let variables = parseVars(template.template) || [];
+
+  // If there isn't one that includes "email", add it
+  // because we need an email to send people verification their letter went through
+  let emailKey = _.find(variables, (v) =>
+    v.toLocaleLowerCase().includes("email")
+  );
+
+  if (!emailKey) {
+    variables = [...variables, "YOUR EMAIL"];
```
[the remark this answers is not in the document]
Comment (blackmad) [reply src/components/letter/LetterForm.tsx:140]: done!

File: src/components/letter/LetterForm.tsx (hunk 1 above)
Comment (blackmad) [reply src/components/letter/LetterForm.tsx:147]: they are their own functions already? I don't really see hooks passed named functions in code - cf the examples in https://reactjs.org/docs/hooks-overview.html or https://reactjs.org/docs/hooks-reference.html#useeffect

basically with hooks, you treat the enclosing function like a class and stuff lives inside it. Pulling these out into their own functions outside the body of the component function would mean passing in all the setX functions which would be a pain.

File: src/components/letter/representative-apis.tsx (hunk 2 above)
Comment (blackmad) [reply src/components/letter/representative-apis.tsx:51]: I'm not sure what you mean / what the benefit would be.

We could have a class that had a constructor that assembled these bits into an OfficialAddress but that's basically what this function does. To make such a constructor safe and easy to use, we'd give it a big named object argument which would look pretty much exactly like what's happening here. I think it would just be extra noise.

It's typechecked all the way through as an OfficalAddress because that's the return type - if this was missing a field from OfficialAddress it would be a compile error.

Comment (blackmad) [reply src/components/letter/representative-apis.tsx:51]: renamed everything to be slightly less terribly named

File: src/components/letter/representative-apis.tsx (hunk 3 above)
Comment (blackmad) [reply src/components/letter/representative-apis.tsx:144]: see last comment

**Comment by blackmad** [comment]
putting @Ceshion's feedback here so I can address it in a bit

(transcribed instead)
1.    use a media-query for field width, horizontally stack the fields unless we're on a small screen
_on hold for now pending bingbong vs ceshion fight to the death over stacked vs aligned mode_
2.    we should consider more consistent language for placeholders (always "your x"?) so that form labels appear consistent
3.    we should also clarify "your email address" so that field could not be confused for a heading
4.    reduce spacing between header and form start, 32 pixels from the header content to the section break, 24 pixels from the section break to the top of the first form label
5.    reduce spacing between form groups, 18-22 pixels as long as it's consistent
6.    set a css variable and use a more fitting color for darkmode, for the background on .bodyWrapper, or remove the background? if not removed, add a border-radius of 4 pixels.
 (image 1)
7.    adjust letter form submit button styling to be consistent with the form fields
 (image 2)
8.    fit success message into the container bounds, revise content to remove personal reference ("Letters Sent! You should get an email in no more than 15 minutes confirming your letters have been sent. If you do not receieve a confirmation, email us at mail-your-rep@blackmad.com"). Is there an easy way with our current setup for the email to be forwarded from defund12's domain so we could use i.e. mail-your-rep@defund12.org?
 (image 3)

**Comment by blackmad** [comment]
1. on hold for now
2. changed all the letter variables to be YOUR X
3. change to be YOUR EMAIL ADDRESS in the code, not changing all the email templates
4. I think this is already 32? It appears identical to me b/w email and letter. Fixed the spacing at the top of the letter form section.
5. padding between form inputs reduced
6. changed the color in both, moved to a var. There appears to be something overall broken with background colors rendering in dark mode - notice that even in prod none of the section backgrounds render - https://defund12.org/nyc?browse
7. adjusted 
image
8. waiting on @bingbongle to verify contact@defund12.org on my sendgrid account

overall looks like this now
image

File: src/components/email-list/EmailList.tsx [hunk 13]
```diff
@@ -2,12 +2,15 @@ import * as React from "react";
 import { v1 as uuid } from "uuid";
 import { StaticQuery, graphql } from "gatsby";
 import { EmailListItem } from "./EmailPageLink";
+import { LetterListItem } from "./LetterListItem";
```
Comment (bingbongle) [inline src/components/email-list/EmailList.tsx:5]: I think we should change this directory structure. let's move relevant files into a components/list directory so that it's no longer solely an `EmailList` and so that `LetterListItem` doesn't live inside of a package called `email-list`

Comment (blackmad) [reply src/components/email-list/EmailList.tsx:5]: agreed. I just ... didn't want to make this commit any larger than it was

File: _emails/us/california/san_francisco.md [hunk 14]
```diff
@@ -30,7 +30,7 @@ recipients:
 - JennyLam@sfusd.edu
 - FaauugaMoliga@sfusd.edu
 - RachelNorton@sfusd.edu
-body: |-
```
Comment (bingbongle) [inline _emails/us/california/san_francisco.md:33]: Why was this removed? It will break this markdown file. Pipe character is needed, or newlines must be explicitly state with \n characters - pipe is the preferred pattern in our case.

Comment (blackmad) [reply _emails/us/california/san_francisco.md:33]: accident? reverted
